### PR TITLE
Remove layer of autogenerated code from executables

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Define executables in a cpp file instead of auto-generating that cpp file. This removes the need for a `Fwd.hpp` file, removes static variables with special names, and makes it (IMO) much easier to understand how an executable is compiled (there's a cpp file with a main function now).

As a side product this PR also makes executable headers compatible with clang-tidy because they don't contain undefined symbols like `metavariables` anymore.

### Upgrade instructions

If you're adding or editing executables in the build system, apply these changes (see one of the existing executables for an example):

- `add_spectre_parallel_executable` was deleted. Use `add_spectre_executable` instead and link `Charmxx::main`.
- Delete the `Fwd.hpp` file.
- Add a cpp file with a main function (specifically, a `CkRegisterMainModule()` function).
- Move the `charm_init_node_funcs` and `charm_init_proc_funcs` to the call to `Parallel::charmxx::register_init_node_and_proc` in the cpp file. Remove the error handling functions from these lists because they are registered automatically.
- If needed, pass template parameters for the metavars from CMake to the cpp file with `target_compile_definitions`.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Elliptic/Executables/Punctures/SolvePunctures.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+extern "C" void CkRegisterMainModule() {
+  Parallel::charmxx::register_main_module<Metavariables>();
+  Parallel::charmxx::register_init_node_and_proc(
+      {&domain::creators::register_derived_with_charm,
+       &register_derived_classes_with_charm<
+           Metavariables::solver::schwarz_smoother::subdomain_solver>,
+       &elliptic::subdomain_preconditioners::register_derived_with_charm,
+       &register_factory_classes_with_charm<Metavariables>},
+      {});
```
Comment (nilsdeppe) [inline src/Elliptic/Executables/Punctures/SolvePunctures.cpp:21]: why not add the init_proc? I just noticed this on this file, but probably happened in others too

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
I did not get through all the files...

File: src/Elliptic/Executables/Punctures/SolvePunctures.cpp (hunk 1 above)
Comment (nilsvu) [reply src/Elliptic/Executables/Punctures/SolvePunctures.cpp:21]: The error handling is added by default (see CharmMain.tpp). We dont have any other proc init functions.

**Review by nilsdeppe — APPROVED** [review]
LGTM, please squash :) Thanks for doing this!

**Pull request merged**